File: python/agentpin/capability.py
"""Capability parsing, matching, and validation for AgentPin."""

import json
from typing import List, Optional, Tuple

from .crypto import sha256_hex


class Capability:
    """A capability in `action:resource` format."""

    def __init__(self, value: str):
        self.value = value

    @staticmethod
    def create(action: str, resource: str) -> "Capability":
        return Capability(f"{action}:{resource}")

    @staticmethod
    def parse(s: str) -> Optional[Tuple[str, str]]:
        idx = s.find(":")
        if idx == -1:
            return None
        return s[:idx], s[idx + 1 :]

    @property
    def action(self) -> Optional[str]:
        parsed = Capability.parse(self.value)
        return parsed[0] if parsed else None

    @property
    def resource(self) -> Optional[str]:
        parsed = Capability.parse(self.value)
        return parsed[1] if parsed else None

    def matches(self, requested: "Capability") -> bool:
        """Check if this capability matches a requested capability.

        Wildcard resources (`*`) match any resource with the same action.
        Scoped resources match if the requested resource starts with the declared resource + '.'.
        """
        self_parsed = Capability.parse(self.value)
        req_parsed = Capability.parse(requested.value)
        if not self_parsed or not req_parsed:
            return False

        self_action, self_resource = self_parsed
        req_action, req_resource = req_parsed

        if self_action != req_action:
            return False
        if self_resource == "*":
            return True
        if self_resource == req_resource:
            return True

        # Scoped matching
        if (
            req_resource.startswith(self_resource)
            and len(req_resource) > len(self_resource)
            and req_resource[len(self_resource)] == "."
        ):
            return True

        return False

    def __str__(self) -> str:
        return self.value

    def __repr__(self) -> str:
        return f"Capability({self.value!r})"

    def __eq__(self, other: object) -> bool:
        if isinstance(other, Capability):
            return self.value == other.value
        return NotImplemented

    def __hash__(self) -> int:
        return hash(self.value)


def capabilities_subset(declared: List[Capability], requested: List[Capability]) -> bool:
    """Check that all requested capabilities are covered by declared capabilities."""
    return all(any(decl.matches(req) for decl in declared) for req in requested)


def capabilities_hash(capabilities: List[Capability]) -> str:
    """Hash capabilities for delegation attestation: SHA-256 of sorted JSON array."""
    sorted_caps = sorted(c.value for c in capabilities)
    json_str = json.dumps(sorted_caps, separators=(",", ":"))
    return sha256_hex(json_str.encode("utf-8"))


CORE_ACTIONS = ["read", "write", "execute", "admin", "delegate"]


def _is_reverse_domain(s: str) -> bool:
    """Check if string looks like a reverse domain prefix (contains a dot)."""
    return "." in s


def validate_capability(cap: Capability) -> None:
    """Validate a capability against the AgentPin taxonomy.

    Raises ValueError if invalid.
    Rules:
    - Must be action:resource format
    - admin:* wildcard rejected
    - Custom (non-core) actions must use reverse-domain prefix
    """
    parsed = Capability.parse(cap.value)
    if not parsed:
        raise ValueError(
            f"Invalid capability format (missing ':'): {cap.value}"
        )
    action, resource = parsed
    if action == "admin" and resource == "*":
        raise ValueError(
            "admin:* wildcard is not allowed; admin capabilities must be explicitly scoped"
        )
    if action in CORE_ACTIONS:
        return
    if not _is_reverse_domain(action):
        raise ValueError(
            f"Custom action '{action}' must use reverse-domain prefix "
            f"(e.g., com.example.{action})"
        )

File: go/pkg/types/capability.go
package types

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"sort"
	"strings"
)

// Capability is an AgentPin capability in the canonical "action:resource"
// format. It serializes as a JSON string for wire compatibility with the
// Rust, JavaScript, and Python SDKs.
type Capability string

// NewCapability constructs a capability from an action/resource pair.
func NewCapability(action, resource string) Capability {
	return Capability(action + ":" + resource)
}

// Action returns the action component of the capability, or "" if the
// capability is malformed.
func (c Capability) Action() string {
	a, _, ok := c.split()
	if !ok {
		return ""
	}
	return a
}

// Resource returns the resource component of the capability, or "" if the
// capability is malformed.
func (c Capability) Resource() string {
	_, r, ok := c.split()
	if !ok {
		return ""
	}
	return r
}

// String returns the canonical "action:resource" representation.
func (c Capability) String() string { return string(c) }

func (c Capability) split() (string, string, bool) {
	idx := strings.IndexByte(string(c), ':')
	if idx < 0 {
		return "", "", false
	}
	return string(c)[:idx], string(c)[idx+1:], true
}

// Matches reports whether a declared capability covers a requested capability.
//
// Wildcard resources ("*") match any resource for the same action. Scoped
// resources match if the requested resource starts with the declared resource
// followed by "." (e.g., "read:codebase" matches
// "read:codebase.github.com/org/repo").
func (c Capability) Matches(requested Capability) bool {
	declAction, declRes, ok := c.split()
	if !ok {
		return false
	}
	reqAction, reqRes, ok := requested.split()
	if !ok {
		return false
	}
	if declAction != reqAction {
		return false
	}
	if declRes == "*" || declRes == reqRes {
		return true
	}
	if strings.HasPrefix(reqRes, declRes) && len(reqRes) > len(declRes) && reqRes[len(declRes)] == '.' {
		return true
	}
	return false
}

// CoreActions lists the AgentPin core action verbs.
var CoreActions = []string{"read", "write", "execute", "admin", "delegate"}

// isReverseDomain reports whether action looks like a reverse-domain prefix
// (e.g., "com.example.scan"). It requires at least two non-empty
// dot-separated segments.
func isReverseDomain(action string) bool {
	parts := strings.Split(action, ".")
	if len(parts) < 2 {
		return false
	}
	for _, p := range parts {
		if p == "" {
			return false
		}
	}
	return true
}

// ValidateCapability validates a capability against the AgentPin taxonomy.
//
//   - Must be in "action:resource" format.
//   - Core actions ("read", "write", "execute", "admin", "delegate") are
//     always valid (with any resource), with one exception:
//   - "admin:*" is rejected — admin must be explicitly scoped.
//   - Custom (non-core) actions MUST use a reverse-domain prefix (e.g.,
//     "com.example.scan:target").
func ValidateCapability(c Capability) error {
	action, resource, ok := c.split()
	if !ok {
		return fmt.Errorf("capability must be in 'action:resource' format")
	}
	if action == "admin" && resource == "*" {
		return fmt.Errorf("admin:* wildcard is not allowed; admin capabilities must be explicitly scoped")
	}
	for _, ca := range CoreActions {
		if ca == action {
			return nil
		}
	}
	if !isReverseDomain(action) {
		return fmt.Errorf("custom action '%s' must use reverse-domain prefix (e.g., com.example.%s)", action, action)
	}
	return nil
}

// CapabilitiesSubset reports whether every requested capability is covered by
// at least one declared capability.
func CapabilitiesSubset(declared, requested []Capability) bool {
	for _, req := range requested {
		matched := false
		for _, decl := range declared {
			if decl.Matches(req) {
				matched = true
				break
			}
		}
		if !matched {
			return false
		}
	}
	return true
}

// CapabilitiesHash hashes a list of capabilities deterministically by sorting
// them alphabetically, JSON-encoding the sorted array, and SHA-256 hashing
// the result. Used by the delegation attestation flow.
func CapabilitiesHash(caps []Capability) string {
	strs := make([]string, len(caps))
	for i, c := range caps {
		strs[i] = string(c)
	}
	sort.Strings(strs)
	b, err := json.Marshal(strs)
	if err != nil {
		// json.Marshal on []string never errors in practice.
		panic(err)
	}
	sum := sha256.Sum256(b)
	return hex.EncodeToString(sum[:])
}
